solidity/contracts/utility/ContractRegistryClient.sol

```diff
@@ -0,0 +1,110 @@
+    bool public allowRegistryUpdate = true; // allow/prevent changing the contract-registry
+
+    /**
+      * @dev initializes a new ContractRegistryClient instance
+      * 
+      * @param  _registry   address of a contract-registry contract
+    */
+    constructor(IContractRegistry _registry) internal validAddress(_registry) {
+        registry = IContractRegistry(_registry);
+        prevRegistry = IContractRegistry(_registry);
+    }
+
+    /**
+      * @dev returns whether or not the caller is authorized to change the contract-registry
+     */
+    function callerIsAuthorized() internal view returns (bool) {
+        return msg.sender == owner;
+    }
+
+    /**
+      * @dev sets the contract-registry to the address of the contract-registry in the contract-registry
+     */
+    function updateRegistry() public {
+        // verify that either updating is allowed or the caller is authorized
+        require(allowRegistryUpdate || callerIsAuthorized());
+
+        // get the address of the contract-registry in the contract-registry
+        address newRegistry = addressOf(CONTRACT_REGISTRY);
+
+        // verify that the new contract-registry is different and not zero
+        require(newRegistry != address(registry) && newRegistry != address(0));
+
+        // set the previous contract-registry as current contract-registry
+        prevRegistry = registry;
+
+        // set the current contract-registry as the new contract-registry
+        registry = IContractRegistry(newRegistry);
+    }
+
+    /**
+      * @dev reverts back to the previous contract-registry
+    */
+    function restoreRegistry() public {
+        // verify that either restoring is allowed or the caller is authorized
+        require(allowRegistryUpdate || callerIsAuthorized());
+
+        // set the current contract-registry as the previous contract-registry
+        registry = prevRegistry;
+
+        // ensure that only an authorized caller can perform future changes
+        allowRegistryUpdate = false;
+    }
+
+    /**
+      * @dev allows/prevents changing the contract-registry
+      * 
+      * @param _disable    true to disable changes, false to enable changes
+    */
+    function disableRegistryUpdate(bool _disable) public ownerOnly {
+        // verify that the caller is authorized
+        require(callerIsAuthorized());
+
+        // allow/prevent changing the contract-registry
+        allowRegistryUpdate = !_disable;
+    }
+
+    /**
+      * @dev returns the address associated with the given contract name
+      * 
+      * @param _contractName    contract name
+      * 
+      * @return contract address
+    */
+    function addressOf(bytes32 _contractName) internal view returns (address) {
+        return registry.addressOf(_contractName);
+    }
+
+    /**
+      * @dev verifies that the caller is mapped to the given contract name
+      * 
+      * @param _contractName    contract name
+    */
+    modifier only(bytes32 _contractName) {

```

Comment:
this and the rest of the code should be at the top (constants -> modifiers -> public functions etc.)

the modifier itself is a nice addition

Reply:
Both `function addressOf(bytes32 _contractName)` and `modifier only(bytes32 _contractName)` are intended for internal use by the different registry clients.
So ideally, I'd like to:
1. Keep them next to each other in the code.
2. Place them after all public functions.
However, while the function is declared `internal`, but the modifier cannot be, purely for technical reasons. Therefore, if we put them both at the top, then we will have an internal function before all public functions.
What do you think?

Reply:
I think that in terms of readability, it's not ideal that you have a list of constants at the end of the file and less important to have them together.
Someone who's inheriting from the contract would need to scroll down to see the constants and it's really not standard (we don't have a header file here...)

Reply:
Yes, the constants can go up.
My argument is with regards to `function addressOf(bytes32 _contractName)` and `modifier only(bytes32 _contractName)`.

Reply:
I would put the modifier last in the list of modifiers and the function last in the list of functions.
Thoughts?

Reply:
That would be the same as what you've originally asked for. Subsequently, the modifier will appear before all public functions and the (internal) function will appear after all public functions.
Proceed?

Reply:
Done